### PR TITLE
shortnames: error if there's no alias and no search registries

### PR DESCRIPTION
**Pull request by vrothberg**
When a short-name could not be resolved to an alias and if there's no
unqualified-search registry, throw an error.

Reported-in: github.com/containers/podman/issues/8573
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@mtrmac @lsm5 @rhatdan PTAL

**Comment by mheon** [comment]
LGTM, thanks @vrothberg

**Comment by lsm5** [comment]
@vrothberg with this change I get: 

```
$ ./bin/podman run --rm idonotexist
Completed short name "idonotexist" with unqualified-search registries (origin: /home/lsm5/.config/containers/registries.conf)
Trying to pull registry.fedoraproject.org/idonotexist:latest...
  manifest unknown: manifest unknown
Trying to pull registry.access.redhat.com/idonotexist:latest...
  name unknown: Repo not found
Trying to pull registry.centos.org/idonotexist:latest...
  manifest unknown: manifest unknown
Trying to pull docker.io/library/idonotexist:latest...
  denied: requested access to the resource is denied
Error: 4 errors occurred while pulling:
 * Error initializing source docker://registry.fedoraproject.org/idonotexist:latest: Error reading manifest latest in registry.fedoraproject.org/idonotexist: manifest unknown: manifest unknown
 * Error initializing source docker://registry.access.redhat.com/idonotexist:latest: Error reading manifest latest in registry.access.redhat.com/idonotexist: name unknown: Repo not found
 * Error initializing source docker://registry.centos.org/idonotexist:latest: Error reading manifest latest in registry.centos.org/idonotexist: manifest unknown: manifest unknown
 * Error initializing source docker://idonotexist:latest: Error reading manifest latest in docker.io/library/idonotexist: errors:
denied: requested access to the resource is denied
unauthorized: authentication required
```

Is that expected output?

**Comment by vrothberg** [comment]
> @vrothberg with this change I get:
> 
> ```
> $ ./bin/podman run --rm idonotexist
> Completed short name "idonotexist" with unqualified-search registries (origin: /home/lsm5/.config/containers/registries.conf)
> Trying to pull registry.fedoraproject.org/idonotexist:latest...
>   manifest unknown: manifest unknown
> Trying to pull registry.access.redhat.com/idonotexist:latest...
>   name unknown: Repo not found
> Trying to pull registry.centos.org/idonotexist:latest...
>   manifest unknown: manifest unknown
> Trying to pull docker.io/library/idonotexist:latest...
>   denied: requested access to the resource is denied
> Error: 4 errors occurred while pulling:
>  * Error initializing source docker://registry.fedoraproject.org/idonotexist:latest: Error reading manifest latest in registry.fedoraproject.org/idonotexist: manifest unknown: manifest unknown
>  * Error initializing source docker://registry.access.redhat.com/idonotexist:latest: Error reading manifest latest in registry.access.redhat.com/idonotexist: name unknown: Repo not found
>  * Error initializing source docker://registry.centos.org/idonotexist:latest: Error reading manifest latest in registry.centos.org/idonotexist: manifest unknown: manifest unknown
>  * Error initializing source docker://idonotexist:latest: Error reading manifest latest in docker.io/library/idonotexist: errors:
> denied: requested access to the resource is denied
> unauthorized: authentication required
> ```
> 
> Is that expected output?

For testing the error case, you need to have `registries.conf` with an empty `unqualified-search-registries` list (or commented out).

**Comment by lsm5** [comment]
> For testing the error case, you need to have `registries.conf` with an empty `unqualified-search-registries` list (or commented out).

Alright, same with `unqualified-search-registries` commented out.

File: pkg/shortnames/shortnames.go [hunk 1]
```diff
@@ -328,21 +328,26 @@ func Resolve(ctx *types.SystemContext, name string) (*Resolved, error) {
...
 		return resolved, nil
 	}
 
 	// If we don't have a TTY, act according to the mode.
 	if !terminal.IsTerminal(int(os.Stdout.Fd())) || !terminal.IsTerminal(int(os.Stdin.Fd())) {
 		switch mode {
 		case types.ShortNameModePermissive:
-			// Permissive falls back to using all candidates.
+			if len(unqualifiedSearchRegistries) == 0 {
+				return nil, errors.Errorf("short-name %q did not resolve to an alias and no unqualified-search registries are defined in %q", name, usrConfig)
```
Comment (mtrmac) [inline pkg/shortnames/shortnames.go:349]: > ```go
> // PullCandidates in the returned value may
> // be empty if there is no matching alias and no unqualified-search registries
> // are configured.
> ```

This is working as documented. If that doesn’t make sense, should we change the API now while it’s early enough that we can pretend it’s not a break?

Comment (vrothberg) [reply pkg/shortnames/shortnames.go:349]: Good catch.  Yes, I think we should change the doc comments. IMHO it's a bug fix.

Comment (mtrmac) [reply pkg/shortnames/shortnames.go:349]: I’d like to go back to the earlier discussions, this was definitely _intentional_ but I can’t remember whether there was a specific rationale.

Implementation-wise, please make this a single check immediately after getting `unqualifiedSearchRegistries`.

Comment (mtrmac) [reply pkg/shortnames/shortnames.go:349]: As far as I can trace this back, https://github.com/containers/image/pull/1061#discussion_r517645860 , which is no longer applicable now that `ResolveLocally` exists.

So, ACK to this change.

Comment (vrothberg) [reply pkg/shortnames/shortnames.go]: Done :heavy_check_mark:

File: pkg/shortnames/shortnames.go [hunk 2]
```diff
@@ -334,15 +337,14 @@ func Resolve(ctx *types.SystemContext, name string) (*Resolved, error) {
...
-	if len(resolved.PullCandidates) <= 1 {
+	if len(resolved.PullCandidates) == 1 {
 		return resolved, nil
 	}
 
 	// If we don't have a TTY, act according to the mode.
 	if !terminal.IsTerminal(int(os.Stdout.Fd())) || !terminal.IsTerminal(int(os.Stdin.Fd())) {
 		switch mode {
 		case types.ShortNameModePermissive:
-			// Permissive falls back to using all candidates.
```
Comment (mtrmac) [inline pkg/shortnames/shortnames.go:345]: Was this removal intentional?

**Review by mtrmac — COMMENTED** [review]
Thanks!

LGTM apart from the two comment nits, please merge without another review round.

File: pkg/shortnames/shortnames.go (hunk 2 above)
Comment (vrothberg) [reply pkg/shortnames/shortnames.go:345]: That was just sloppy me. Thanks for catching!
